presto-common/src/main/java/com/facebook/presto/common/zorder/ZOrderIntegers.java

```diff
@@ -0,0 +1,124 @@
+                    input.size(), mostSignificantSetBitPositions.size()));
+        }
+
+        for (int index = 0; index < input.size(); index++) {
+            if (input.get(index) < 0) {
+                throw new IllegalArgumentException("Our current Z-Ordering implementation does not support negative input numbers.");
+            }
+        }
+    }
+
+    // TODO: make int and long z-address version
+    public byte[] encode(List input)
+            throws IllegalArgumentException
+    {
+        checkInputValidity(input);
+
+        int totalBitLength = 0;
+        int maxBitLength = 0;
+        for (int bitLength : mostSignificantSetBitPositions) {
+            totalBitLength += bitLength;
+            maxBitLength = Math.max(maxBitLength, bitLength);
+        }
+
+        byte[] address = new byte[(totalBitLength + 7) >> 3];
+        int bitIndex = totalBitLength - 1;
+        int bitOffset = bitIndex & 7;
+        // Interweave input bits into address from the most significant bit to preserve data locality
+        for (int inputBitPos = 0; inputBitPos < maxBitLength; inputBitPos++) {
+            for (int index = 0; index < input.size(); index++) {
+                if (inputBitPos >= mostSignificantSetBitPositions.get(index)) {
+                    continue;
+                }
+                int maskedBit = ((Integer) input.get(index) >> (mostSignificantSetBitPositions.get(index) - inputBitPos - 1)) & 1;
+                address[bitIndex >> 3] |= maskedBit << bitOffset;
+                bitIndex--;
+                bitOffset = bitIndex & 7;
+            }
+        }
+        return address;
+    }
+
+    public List decode(byte[] address)
+    {
+        int listLength = mostSignificantSetBitPositions.size();
+        int[] output = new int[listLength];
+
+        int totalBitLength = 0;
+        int maxBitLength = 0;
+        for (int bitLength : mostSignificantSetBitPositions) {
+            totalBitLength += bitLength;
+            maxBitLength = Math.max(maxBitLength, bitLength);
+        }
+
+
+        int bitIndex = totalBitLength - 1;
+        int bitOffset = bitIndex & 7;
+        int outputBitPos = 0;
+        // Un-weave address into original integers
+        while (bitIndex > 0) {
+            for (int index = 0; index < listLength; index++) {
+                if (outputBitPos >= mostSignificantSetBitPositions.get(index)) {
+                    continue;
+                }
+                int maskedBit = (address[bitIndex >> 3] >> bitOffset) & 1;
+                output[index] |= maskedBit << (mostSignificantSetBitPositions.get(index) - outputBitPos - 1);
+                bitIndex--;
+                bitOffset = bitIndex & 7;
+            }
+            outputBitPos++;
+        }
+
+        List<Integer> intList = new ArrayList<>(output.length);
+        for (int x : output) {
+            intList.add(x);
+        }
```

Comment:
Sounds there are other nicer methods to convert int[] to List<Integer>: https://stackoverflow.com/questions/1073919/how-to-convert-int-into-listinteger-in-java